DESIGN.md

```diff
@@ -0,0 +1,756 @@
+drop(buf);
+```
+
+The runtime supports initializing multiple pools containing buffers of different
+sizes. When submitting a read operation, the caller specifies the pool from
+which to draw a buffer.
+
+```rust
+// Allocate a buffer pool
+let my_pool = BufferPool::builder()
+    .buffer_size(16_384)
+    .num_buffers(256)
+    .build();
+
+// Create the runtime
+let mut rt = tokio_uring::runtime::Runtime::new()?;
+
+// Provide the buffer pool to the kernel. This passes
+// ownership of the pool to the kernel.
+let pool_token = rt.provide_buffers(MY_POOL, 256, 4 * 4096)?;
+    
+rt.block_on(async {
+    // ...
+    let buf = file.read_at_prepared(0, 1024, pool_token).await?;
+});
+```
+
+## Buffer management
+
+Buffers passed to read and write operations must remain alive and pinned to a
+memory location while operations are in-flight, ruling out `&mut [u8]` as an
+option. Additionally, read and write operations may reference buffers using a
+pointer or, when the buffer is part of a pre-registered buffer pool, using a
+numeric buffer identifier. 
+
+The tokio-uring crate will provide its buffer type, `IoBuf`, to use when reading
+and writing.
+
+```rust
+pub struct IoBuf {
+    kind: Kind,
+}
+
+enum Kind {
+    /// A vector-backed buffer
+    Vec(Vec<u8>),
+
+    /// Buffer pool backed buffer. The pool is managed by io-uring.
+    Provided(ProvidedBuf),
+}
+```
+
+Internally, the IoBuf type is either backed by individually heap-allocated
+memory or a buffer pool entry. Acquiring an `IoBuf` is done via
+`IoBuf::with_capacity` or checking-out an entry from a pool. 
+
+```rust
+// Individually heap-allocated
+let my_buf = IoBuf::with_capacity(4096);
+
+// Checked-out from a pool
+match my_buffer_pool.checkout() {
+    Ok(io_buf) => ...,
+    Err(e) => panic!("buffer pool empty"),
+}
+```
+
+On drop, if the `IoBuf` is a buffer pool member, it is checked back in. If the
+kernel initially checked out the buffer as part of a read operation, an io-uring
+operation is issued to return it. Submitting the io-uring operation requires the
+buffer to remain on the same thread that checked it out and is enforced by
+making the `IoBuf` type `!Send`. 
+
+`IoBuf` provides an owned slice API allowing the caller to read to and write from
+a buffer's sub-ranges. 
+
+```rust
+pub struct Slice {
+    buf: IoBuf,
+    begin: usize,
+    end: usize,
+}
+
+impl IoBuf {
+    fn slice(self, range: impl ops::RangeBounds<usize>) -> Slice { .. }
+}
+
+// Write a sub-slice of a buffer
+my_file.write(my_io_buf.slice(10..20)).await
+```
+
+A slice end may go past the buffer's length but not past the capacity, enabling
+reads to uninitialized memory.
+
+```rust
+// The buffer's memory is uninitialized
+let mut buf = IoBuf::with_capacity(4096);
+let slice = buf.slice(0..100);
+
+assert_eq!(slice.len(), 0);
+assert_eq!(slice.capacity(), 100);
+
+// Read data from a file into the buffer
+let BufResult(res, slice) = my_file.read_at(slice, 0);
+
+assert_eq!(slice.len(), 100);
```

Comment:
This is nitpicking, but this assert seems overly aggressive given that the file may contains fewer than 100 bytes.

Reply:
I "know" the file contains more than 100 bytes :laugh: